### PR TITLE
Support block volume mode PVC

### PR DESCRIPTION
**Pull request by dkeven**
Signed-off-by: dkeven <keven@kubesphere.io>

**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

This PR adds suppport for PVCs whose `volumeMode` is `block`, which is not supported before.

### CODE REVIEW

File: pkg/disk/rpcserver/nodeserver.go [hunk 1]
```diff
@@ -525,3 +556,37 @@ func (ns *NodeServer) getBlockSizeBytes(devicePath string) (int64, error) {
...
+	if err != nil {
+		klog.Errorf("Failed to determine whether mountPath:%s exists: %v", mountPath, err)
+		return err
+	}
+	if exists {
+		return nil
+	}
+
+	if isBlockMode {
+		//#nosec
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/nodeserver.go:574]: what's the meaning of this comment?

File: pkg/disk/rpcserver/nodeserver.go [hunk 2]
```diff
@@ -525,3 +556,37 @@ func (ns *NodeServer) getBlockSizeBytes(devicePath string) (int64, error) {
...
+			klog.Errorf("Failed to create mountPath:%s with error: %v", mountPath, err)
+			return err
+		}
+		if err = pathFile.Close(); err != nil {
+			klog.Errorf("Failed to close mountPath:%s with error: %v", mountPath, err)
+			return err
+		}
+	} else {
+		// Create a directory
+		if err = os.MkdirAll(mountPath, 0750); err != nil {
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/nodeserver.go:586]: should be `if err := os.MkdirAll`

File: pkg/disk/rpcserver/nodeserver.go [hunk 3]
```diff
@@ -525,3 +556,37 @@ func (ns *NodeServer) getBlockSizeBytes(devicePath string) (int64, error) {
...
+
+	if isBlockMode {
+		//#nosec
+		pathFile, err := os.OpenFile(mountPath, os.O_CREATE|os.O_RDWR, 0750)
+		if err != nil {
+			klog.Errorf("Failed to create mountPath:%s with error: %v", mountPath, err)
+			return err
+		}
+		if err = pathFile.Close(); err != nil {
+			klog.Errorf("Failed to close mountPath:%s with error: %v", mountPath, err)
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/nodeserver.go:581]: almost all errors in this function were logged. This is not recommended. Should return the error, and let the caller decide whether to log the error.

File: pkg/disk/rpcserver/controllerserver.go [hunk 4]
```diff
@@ -676,13 +676,17 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
 	if err != nil {
 		return nil, status.Errorf(codes.OutOfRange, err.Error())
 	}
+
+	// If volume is block mode, it doesn't need NodeExpandVolume operation
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/controllerserver.go:680]: should be `If volume is in block mode`, and as all followings.

File: pkg/disk/rpcserver/nodeserver.go [hunk 5]
```diff
@@ -91,47 +92,44 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	stagePath := req.GetStagingTargetPath()
 	volumeId := req.GetVolumeId()
 
+	// check if volume exists in the first place
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/nodeserver.go:95]: remove `in the first place` as it's a bit misleading,  this is not the first check in this function.

File: pkg/disk/rpcserver/nodeserver.go [hunk 6]
```diff
@@ -224,6 +222,15 @@ func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	// set parameter
 	volumeId := req.GetVolumeId()
 	targetPath := req.GetStagingTargetPath()
+
+	// according to the CSI spec, the CO shall ensure this path is a directory
+	// which is incompatible with a block device
+	// we skip the staging operation in this case
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/nodeserver.go:228]: should just say `skip staging if volume in block mode`

File: pkg/disk/rpcserver/nodeserver.go [hunk 7]
```diff
@@ -488,7 +497,29 @@ func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context,
 	}
 
 	volumePath := req.GetVolumePath()
+	// block mode volume's stats can't be retrieved like that of fs mode ones
```
Comment (stoneshi-yunify) [inline pkg/disk/rpcserver/nodeserver.go:500]: should be `... like those filesystem volumes`

**Review by stoneshi-yunify — CHANGES_REQUESTED** [review]
see my comments. Make sure new code does not break old FS volume logic.

**Comment by dkeven** [comment]
@stoneshi-yunify Thanks for your detailed review, I've updated the code as suggested.

> see my comments. Make sure new code does not break old FS volume logic.

I've tested both the FS and Block mode, and they works correctly.

> what's the meaning of this comment?

Can't reply in the thread as new commit's been force-pushed.

Some linters, like gosec, don't allow creating file whose path is a variable, this comment disables the check.

As we don't do this kind of linting, I think it's okay to remove this comment.

**Review by stoneshi-yunify — APPROVED** [review]
/lgtm

**Pull request merged**